### PR TITLE
Start sandbox logic shared for demo/sandbox  cmds #minor

### PR DESCRIPTION
**Pull request by dav009**
# TL;DR
Logic for start sandbox and demo is redudant. This PR unifies start logic under a pkg/sandbox package

## Type
- [ ] Bug Fix
- [x] Feature
- [ ] Plugin

## Are all requirements met?

- [ ] Code completed
- [ ] Smoke tested
- [x] Unit tests added
- [ ] Code documentation added
- [ ] Any pending items have an associated Issue

## Complete description
As tracking issue describes there is a lot of repeated logic between `sandbox` and `demo` commands. This PR unifies the logic for starting a sandbox under a new package `pkg/sandbox`. Upcoming Prs will address `teardown` and other subcommands.

This PR:

- Unifies logic for starting demo and sandbox
-  Introduces a new package `pkg/sandbox`
- Starting a demo is starting a sandbox with: (i) a particular docker image, (ii) prime pods running.
- `Config` for `sandbox` now lives in `pkg/sandbox`
- `ImagePullPolicy` and `ImagePullOptions` are part of `pkg/docker`
- Most of this PR is just moving code . 
- Tests are mostly kept as they were
- added `StartDemoCluster` and `StartSandboxCluster` which rely on calling `StartCluster` with needed params for each particular case

⚠️  I still have to create/destroy and check both demo and sandboxes runs locally. Tests and lints are passing at the moment.

Whats next?

In other batch of work/PR:
- do the same for teardown
- do the same for status

## Tracking Issue
https://github.com/flyteorg/flyte/issues/2472

### CODE REVIEW

**Review by pmahindrakar-oss — COMMENTED** [review]
LGTM . Please update on the test results and we can get this checkedin.

File: pkg/sandbox/config.go [hunk 1]
```diff
@@ -0,0 +1,35 @@
+package sandbox
+
+import "github.com/flyteorg/flytectl/pkg/docker"
+
+//Config holds configuration flags for sandbox command.
+type Config struct {
```
[the remark this answers is not in the document]
Comment (dav009) [reply pkg/sandbox/config.go:6]: @pmahindrakar-oss my motivation for moving this struct here is that I made this part of the inputs to some of the functions in `pgk/sandbox`. As shown in the signatures below

```golang
func StartCluster(ctx context.Context, args []string, sandboxConfig *Config, primePod bool, defaultImageName string)
func StartDemoCluster(ctx context.Context, args []string, sandboxConfig *Config)
```
what do you think it is better?
- just pass all of this fields directly into these functions
- define this config in config/subcommand and then import it in pkg/sanbox?  ( I wonder if this might cause an import loop)
- create another struct essentially with the same fields. Maybe we can embed them?
```golang
// lives in cmd/config/subcommand
type CLIConfig struct {
   // sandbox.Config
   Config
}
```

[the remark this answers is not in the document]
Comment (pmahindrakar-oss) [reply pkg/sandbox/config.go:6]: Option 2 & 3 should be good. 
For the option 2 . I dont see an import loop issue since 

subcommand is only imported in cmd module
pkg is imported in cmd module
here you are introducing a dependency from subcommand -> pkg which should be fine.

Option 3 might lead to issue with pflag generation though as embedded struct may not be supported but you can check.

**Comment by dav009** [comment]
@pmahindrakar-oss 
- moved config to cmd/config
- ran `flytectl demo start` and `flytectl sandbox start` locally succesfully. 
- Fixed a bunch of issues that I missed on the way:
  - essentially passing different ports (yeah, ports are different for demo and sandbox)
  - different way to look for a fully qualified docker image 

codecov looks very sad now. Probably we could have another `good first issue` to make `StartSandboxCluster` and `StartDemoCluster` better testable, it is hard to mock things and verify function calls.
I was thinking on a struct/interface like: 
```
type Sandbox struct{
  start : func(...)
  teardown: func(...)
}
```
and then we would be able to mock and tests what values other functions inject. i.e: Starting a demo cluster means using `demoDockerImage`, `demoPorts`

**Comment by pmahindrakar-oss** [comment]
Yes we can do the tests as followup.Thanks for making this change

**Review by pmahindrakar-oss — APPROVED** [review]
